plugins/processors/ifname/ifname.go

```diff
@@ -0,0 +1,328 @@
+	ifXTable *si.Table `toml:"-"`
+
+	rwLock sync.RWMutex `toml:"-"`
+	cache  *LRUCache    `toml:"-"`
+
+	parallel parallel.Parallel    `toml:"-"`
+	acc      telegraf.Accumulator `toml:"-"`
+}
+
+type nameMap map[uint64]string
+
+func (d *IfName) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *IfName) Description() string {
+	return "Add a tag of the network interface name looked up over SNMP by interface number"
+}
+
+func (h *IfName) Init() error {
+	var err error
+	h.ifTable, err = makeTable("IF-MIB::ifTable")
+	if err != nil {
+		return err
+	}
+	h.ifXTable, err = makeTable("IF-MIB::ifXTable")
+	if err != nil {
+		return err
+	}
+
+	c := NewLRUCache(h.CacheSize)
+	h.cache = &c
+
+	return nil
+}
+
+func (d *IfName) addTag(metric telegraf.Metric) {
+	//this is called by the pool but locking is handled inside getMap
+	// if metric == nil {
+	// 	return
+	// }
+
+	agent, ok := metric.GetTag(d.AgentTag)
+	if !ok {
+		//agent tag missing
+		return
+	}
+
+	num_s, ok := metric.GetTag(d.SourceTag)
+	if !ok {
+		//source tag missing
+		return
+	}
+
+	num, err := strconv.ParseUint(num_s, 10, 64)
+	if err != nil {
+		//source tag not a uint
+		d.Log.Infof("couldn't parse source tag as uint")
+		return
+	}
+
+	m, err := d.getMap(agent)
+	if err != nil {
+		d.Log.Infof("couldn't retrieve the table of interface names: %w", err)
+		return
+	}
+
+	name, found := (*m)[num]
+	if !found {
+		//interface num isn't in table
+		d.Log.Infof("interface number %d isn't in the table of interface names", num)
+		return
+	}
+	metric.AddTag(d.DestTag, name)
+}
+
+func (d *IfName) Apply(metrics ...telegraf.Metric) []telegraf.Metric {
+	for _, metric := range metrics {
+		d.addTag(metric)
+	}
+	return metrics
+}
+
+func (d *IfName) Start(acc telegraf.Accumulator) error {
+	d.acc = acc
+
+	fn := func(m telegraf.Metric) []telegraf.Metric {
+		d.addTag(m)
+		return []telegraf.Metric{m}
+	}
+
+	if d.Ordered {
+		d.parallel = parallel.NewOrdered(acc, fn, 10000, d.MaxParallelLookups)
+	} else {
+		d.parallel = parallel.NewUnordered(acc, fn, d.MaxParallelLookups)
+	}
+	return nil
+}
+
+func (d *IfName) Add(metric telegraf.Metric, acc telegraf.Accumulator) error {
+	d.parallel.Enqueue(metric)
+	return nil
+}
+
+func (d *IfName) Stop() error {
+	d.parallel.Stop()
+	return nil
+}
+
+// getMap gets the interface names map either from cache or from the SNMP
+// agent
+func (d *IfName) getMap(agent string) (*nameMap, error) {
+	// Check cache first
+	d.rwLock.RLock()
+	m, ok := d.cache.Get(agent)
+	d.rwLock.RUnlock()
+	if ok {
+		return m, nil
+	}
+
+	// The cache missed so make an SNMP request.  Write the response
+	// to cache and return it
+	d.rwLock.Lock()
+	defer d.rwLock.Unlock()
+
+	// Check cache again while holding write lock to prevent duplicate
+	// SNMP requests.
+	var err error
+	m, ok = d.cache.Get(agent)
+	if ok {
+		return m, nil
+	}
+
+	m, err = d.getMapRemote(agent)
+	if err != nil {
+		return nil, err
+	}
+
+	d.cache.Put(agent, m)
```

Comment:
Can these maps be updated if the interfaces change?  Maybe something for later.

Reply:
Interfaces change so rarely I didn't think it was important to add a way to invalidate and reload the maps in the first version.  I agree it would be a good future change.

Reply:
Virtual interfaces get added and removed all the time. Even frequently for physical ones…

Reply:
Hi @Hipska, do you have a suggestion on how to handle caching of interface names? We can't get the names from the agent every time they're needed because it's a relatively slow network transaction. We also don't want to cache the names forever because they change. Would it be sufficient to expire cached names after a set amount of time, configurable in the plugin settings?  Maybe we should also invalidate the cache entry when a request comes in for an interface number that isn't in the cached list?

Reply:
Yeah, a timeout based on the polling interval is what I would expect, that also handles the invalidate problem when a new interface is found. Otherwise indeed invalidate the cache if number of interfaces change and also after some timeout.

Reply:
Since ifname is a processor it doesn't have a polling interval.  Metrics will come into ifname from various inputs (snmp, sflow, possibly others) each of which could have different intervals.  The processor api also doesn't have a way to determine which input a given metric came from.  These limitations mean it's more practical to add a new setting for cache time to live.

I'm working on changes to do this.  I'll also have it invalidate the cache entry if the interface number isn't in the cached list.